File: sdk/python/costorah/reliability/connection_pool.py
"""
ConnectionPool — a single shared `httpx.AsyncClient`, reused across every
delivery attempt instead of opening a new connection per request. httpx
already pools/keeps-alive connections per host internally; this wrapper's
job is just to make sure the whole reliability layer shares exactly one
such client (and its pool) rather than constructing one per request.
"""

from __future__ import annotations

import httpx

from costorah.config import Config
from costorah.version import __version__


class ConnectionPool:
    def __init__(
        self,
        config: Config,
        *,
        max_connections: int = 20,
        max_keepalive_connections: int = 10,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self._config = config
        self.requests_sent = 0
        self._client = httpx.AsyncClient(
            base_url=config.endpoint,
            timeout=config.timeout,
            verify=config.verify_tls,
            transport=transport,
            limits=httpx.Limits(
                max_connections=max_connections,
                max_keepalive_connections=max_keepalive_connections,
            ),
            headers={
                "Authorization": f"Bearer {config.api_key}",
                "User-Agent": f"costorah-python/{__version__}",
            },
        )

    async def post(self, path: str, *, content: bytes, headers: dict[str, str]) -> httpx.Response:
        self.requests_sent += 1
        return await self._client.post(path, content=content, headers=headers)

    async def aclose(self) -> None:
        await self._client.aclose()

File: sdk/javascript/src/reliability/connectionPool.ts
/**
 * ConnectionPool — reuses HTTP connections across delivery attempts
 * instead of opening a new one per request.
 *
 * Node's global `fetch` (and every `FetchLike` implementation this SDK
 * accepts) is backed by undici, which already keep-alives and pools
 * connections per origin automatically — there is no separate "open a
 * pool" step the way `httpx.AsyncClient` needs in Python. This class's
 * job is just to make sure the whole reliability layer issues every
 * request through the *same* fetch implementation/instance (so whatever
 * pooling the runtime does is actually shared), plus track basic stats.
 */

import type { FetchLike } from "../http.js";
import type { ResolvedConfig } from "../config.js";
import { VERSION } from "../version.js";

export class ConnectionPool {
  private readonly config: ResolvedConfig;
  private readonly fetchImpl: FetchLike;
  requestsSent = 0;

  constructor(config: ResolvedConfig, fetchImpl?: FetchLike) {
    this.config = config;
    this.fetchImpl = fetchImpl ?? globalThis.fetch;
  }

  async post(
    path: string,
    body: Uint8Array,
    headers: Record<string, string>,
  ): Promise<Response> {
    this.requestsSent += 1;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.config.timeout * 1000);
    try {
      return await this.fetchImpl(`${this.config.endpoint}${path}`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.config.apiKey}`,
          "User-Agent": `costorah-js/${VERSION}`,
          ...headers,
        },
        body: body as unknown as BodyInit,
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timer);
    }
  }
}
